### PR TITLE
[ENH] Peformance improvement for `expand_grid`

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Dump down into numpy arrays where possible
- Build a dataframe, as opposed to `pd.concat` for speed/some memory savings
- cascades to `complete`;
- patch for `complete` (sort=False is more efficient; `explicit` logic more efficient as well)

```py
# take tests with a pinch of salt
df1 = pd.DataFrame({"x":range(1,3), "y":[2,1]})
df1 = pd.concat([df1]*100)
df2 = pd.DataFrame({"x":[1,2,3],"y":[3,2,1]})
df2 = pd.concat([df2]*100)
df3 = pd.DataFrame({"x":[2,3],"y":["a","b"]})
df3 = pd.concat([df3]*100)

data = {"df1":df1, "df2":df2, "df3":df3}

------> dev
 %timeit jn.expand_grid(others = data) 
3.49 s ± 260 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%memit jn.expand_grid(others = data)
peak memory: 1936.24 MiB, increment: 1716.80 MiB


------> PR
 %timeit jn.expand_grid(others = data)
620 ms ± 58.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%memit jn.expand_grid(others = data)
peak memory: 905.20 MiB, increment: 763.66 MiB
```

**This PR improves `expand_grid`.**

# PR Checklist

Please ensure that you have done the following:

1. [ ] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [ ] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1033--pyjanitor.netlify.app

File: janitor/utils.py [hunk 1]
```diff
@@ -42,99 +43,117 @@ def check(varname: str, value, expected_types: list):
...
-
     if value.ndim > 2:
         raise ValueError(
-            """
-            expand_grid works only
-            on 1D and 2D arrays.
-            """
+            "expand_grid works only on 1D and 2D arrays. "
+            "The provided array for {key} however "
+            f" has a dimension of {value.ndim}."
```
Comment (thatlittleboy) [inline janitor/utils.py]:
```suggestion
            "expand_grid works only on 1D and 2D arrays. "
            f"The provided array for {key} however "
            f"has a dimension of {value.ndim}."
```

File: janitor/utils.py [hunk 2]
```diff
@@ -42,99 +43,117 @@ def check(varname: str, value, expected_types: list):
...
+            arr = extract_array(arr, extract_numpy=True)[grid_index]
+            if not name:
+                name = num
+                num += 1
+            contents[(key, name)] = arr
+        return contents
+    name = value.name
+    if not name:
+        name = 0
+    return {(key, name): extract_array(value, extract_numpy=True)[grid_index]}
```
Comment (thatlittleboy) [inline janitor/utils.py]: It seems to me that these two should be separate dispatch functions (one for `pd.Index`, one for `pd.MultiIndex`)

File: midpoint.csv [hunk 3]
```diff
@@ -0,0 +1,3 @@
+,a_new,b
```
Comment (thatlittleboy) [inline midpoint.csv]: Remove this file

File: tests/functions/test_expand_grid.py [hunk 4]
```diff
@@ -181,6 +181,22 @@ def test_Index(df):
     assert_frame_equal(result, expected)
 
 
+@settings(deadline=None)
+@given(df=df_strategy())
+def test_Index_name_none(df):
+    """Test expand_grid output"""
```
Comment (thatlittleboy) [inline tests/functions/test_expand_grid.py]: One small request from me to be a bit more descriptive on what this test does; from the test name, it seems it is specifically testing when value is an Index, with no name?

File: tests/functions/test_expand_grid.py [hunk 5]
```diff
@@ -200,6 +216,25 @@ def test_MultiIndex(df):
     assert_frame_equal(result, expected)
 
 
+@settings(deadline=None)
+@given(df=categoricaldf_strategy())
+def test_MultiIndex_names_none(df):
+    """Test expand_grid output"""
```
Comment (thatlittleboy) [inline tests/functions/test_expand_grid.py]: Similarly for this, please. (more descriptive test docstring)

**Review by thatlittleboy — CHANGES_REQUESTED** [review]
I just did a partial scan-through of the PR, would be great if you address these initial comments first. Will look through a bit more later on.

File: janitor/utils.py (hunk 2 above)
Comment (samukweku) [reply janitor/utils.py]: The thing is pd.Index is the superset, so any dispatch to multiindex will go instead to the pd.index and error out. Open to suggestions on how to go about it

Comment (thatlittleboy) [reply janitor/utils.py]: Could you give me an example? I can't quite reproduce it.

```python
import pandas as pd
from functools import singledispatch

@singledispatch
def foo(x):
    raise TypeError

@foo.register(pd.MultiIndex)
def _foo(x):
    print("I'm in multiindex dispatch")

@foo.register(pd.Index)
def _foo(x):
    print("I'm in index dispatch")


m = pd.Index([1, 2, 3])
foo(m)  # dispatch to index
m = pd.MultiIndex.from_tuples([(1, 1), (2, 2), (3, 3)])
foo(m)  # dispatch to multindex
```
image

File: janitor/functions/complete.py [hunk 6]
```diff
@@ -212,112 +212,141 @@ def _computations_complete(
...
+    return uniques
 
 
 @functools.singledispatch
-def _complete_column(column: str, df):
+def _complete_column(column: str, df, sort):
     """
     Args:
         column : str/list/dict
         df: Pandas DataFrame
```
Comment (thatlittleboy) [inline janitor/functions/complete.py:329]: Let's also document down the newly-added `sort` parameter.

File: janitor/functions/complete.py [hunk 7]
```diff
@@ -212,112 +212,141 @@ def _computations_complete(
...
+    # dirty tests also show that drop_duplicates
+    # is faster than pd.unique for fairly large data
+
     column = df[column]
+    dupes = column.duplicated()
+
+    if dupes.any():
+        column = column[~dupes]
+
+    if sort and (column.is_monotonic_increasing is False):
```
Comment (thatlittleboy) [inline janitor/functions/complete.py]:
```suggestion
    if sort and not column.is_monotonic_increasing:
```

File: janitor/functions/complete.py [hunk 8]
```diff
@@ -371,8 +404,13 @@ def _sub_complete_column(column, df):  # noqa: F811
...
             arr = pd.Series(arr)
 
-        if not arr.is_unique:
-            arr = arr.drop_duplicates()
+        dupes = arr.duplicated()
+
+        if dupes.any():
+            arr = arr[~dupes]
+
+        if sort and (arr.is_monotonic_increasing is False):
```
Comment (thatlittleboy) [inline janitor/functions/complete.py]:
```suggestion
        if sort and not arr.is_monotonic_increasing:
```

File: janitor/functions/utils.py [hunk 9]
```diff
@@ -142,14 +141,12 @@ def patterns(regex_pattern: Union[str, Pattern]) -> Pattern:
...
-    to expand each input to the length of the cumulative product of
-    all inputs in `others`.
+    Uses numpy's `mgrid` to generate indices, which is used to
+    `explode` all the inputs in `others`.
 
     There is a performance penalty for small entries
-    (lenght less than 10)
     in using this method, instead of `itertools.product`;
-    however, there are significant performance benefits
+    however, there are  significant performance benefits
```
Comment (thatlittleboy) [inline janitor/functions/utils.py]: Unnecessary change here

File: janitor/functions/utils.py [hunk 10]
```diff
@@ -167,13 +164,11 @@ def _computations_expand_grid(others: dict) -> pd.DataFrame:
...
         if is_scalar(value):
-            value = pd.DataFrame([value])
-        elif (not isinstance(value, pd.Series)) and is_extension_array_dtype(
-            value
-        ):
-            value = pd.DataFrame(value)
+            value = np.asarray([value])
         elif is_list_like(value) and (not hasattr(value, "shape")):
             value = np.asarray([*value])
+        if value.size == 0:
```
Comment (thatlittleboy) [inline janitor/functions/utils.py]:
```suggestion
        if not value.size:
```

File: janitor/functions/complete.py [hunk 11]
```diff
@@ -212,112 +212,141 @@ def _computations_complete(
...
-    # and merge back to the main dataframe
 
-    # to get a name that does not exist in the columns
-    indicator = "".join(df.columns)
-    trimmed = df.loc(axis=1)[column_checker]
-    uniques = uniques.merge(
-        trimmed, how="left", sort=False, copy=False, indicator=indicator
+    columns = df.columns
+    indicator = False
+    if (fill_value is not None) and (explicit is False):
```
Comment (thatlittleboy) [inline janitor/functions/complete.py]: Since explicit is already a boolean, we can just do `not explicit` instead.

```suggestion
    if fill_value is not None and not explicit:
```

File: janitor/functions/complete.py [hunk 12]
```diff
@@ -212,112 +212,141 @@ def _computations_complete(
...
-    Returns a Series or DataFrame, with no duplicates.
+    Returns a DataFrame, with no duplicates.
     """
     if all_strings:
-        uniques = {col: df[col].unique() for col in columns}
+        if sort:
+            uniques = {}
+            for col in columns:
+                column = extract_array(df[col], extract_numpy=True)
+                _, column = pd.factorize(column, sort=sort)
```
Comment (thatlittleboy) [inline janitor/functions/complete.py:294]: Is there any value to extracting the array ourselves here rather than leaving it to pandas?
`pd.factorize` / `pd.Series.factorize` seems to already [do this internally](https://github.com/pandas-dev/pandas/blob/06d230151e6f18fdb8139d09abf539867a8cd481/pandas/core/algorithms.py#L725).

**Review by thatlittleboy — CHANGES_REQUESTED** [review]

File: janitor/functions/complete.py (hunk 7 above)
Comment (samukweku) [reply janitor/functions/complete.py]: They do d same thing, and I find the first one more explicit (I am guilty of using them interchangeably tho)

Comment (thatlittleboy) [reply janitor/functions/complete.py]: Yeah, they do.
The second is more widely regarded to be pythonic, though. (`is False` is really only used if there is an explicit need to check for False rather than generic Falsy values like 0, None, `[]`, etc.)

image
https://www.python.org/dev/peps/pep-0008/

File: janitor/functions/complete.py (hunk 12 above)
Comment (samukweku) [reply janitor/functions/complete.py:294]: Only if it is already an array; any pandas object is converted to an Index, except Categorical, which returns a categorical.

```py
In [2]: box = ['b', 'b', 'a', 'c', 'b']

In [3]: box = pd.Series(box)

In [4]: pd.factorize(box)
Out[4]: (array([0, 0, 1, 2, 0]), Index(['b', 'a', 'c'], dtype='object'))

In [5]: pd.factorize(box)[-1]
Out[5]: Index(['b', 'a', 'c'], dtype='object')
```

**Comment by samukweku** [comment]
any ideas @thatlittleboy regarding this error message : ``ERROR    -  Error reading page 'api/biology.md': No module named 'mkdocstrings.handlers.python'`` - this is why the build documentation is failing @ericmjl

**Comment by thatlittleboy** [comment]
@samukweku Yeah, it's because `mkdocstrings` just got updated to 0.18 on conda-forge, which added a new optional dependency for the legacy python handler (it got refactored out into its own project);
and the PR for it to be added into conda-forge hasn't been merged in yet.

You can fix the version for `mkdocstrings` in `environment-dev.yml` to be <0.18.0 for the time being; It should fix the issue.
I'll push up a PR fix for compatibility against mkdocstrings >=0.18.0 when I find the time.

**Comment by samukweku** [comment]
@thatlittleboy thanks ... you seem to have a finger on infrastructure ... looking forward to learning more from u

**Review by thatlittleboy — APPROVED** [review]

**Review by ericmjl — APPROVED** [review]
@samukweku and @thatlittleboy, thank you to both of you here!

**Comment by ericmjl** [comment]
@thatlittleboy please do the honours!
